Translation of Strict Mode (#121)
reactjs/uk.reactjs.org: content/docs/strict-mode.md

---
id: strict-mode
title: Суворий режим
permalink: docs/strict-mode.html
---

> `StrictMode` – це інструмент для виявлення потенційних проблем у додатку. `StrictMode` не рендерить видимого UI так само як `Fragment`. Він активує додаткові перевірки та попередження для своїх нащадків.

Reviewer: ```suggestion
`StrictMode` – це інструмент для виявлення потенційних проблем у додатку. Так само як і `Fragment`, `StrictMode` не рендерить видимого UI. Він активує додаткові перевірки та попередження для своїх нащадків.
```

> Примітка:
>
> Перевірки суворого режиму виконуються лише в режимі розробки; _вони не впливають на продакшн-збірку_.

Ви можете увімкнути суворий режим для будь-якої частини вашого додатку. Наприклад:
`embed:strict-mode/enabling-strict-mode.js`

> У наведеному вище прикладі перевірки суворого режиму *не* будуть виконуватись для компонентів `Header` та `Footer`. Проте компоненти `ComponentOne` та `ComponentTwo`, а також всі їхні нащадки матимуть перевірки.

Reviewer: `У наведеному вище прикладі,` (кома)

`StrictMode` наразі допомагає в:
* [Ідентифікації компонентів з небезпечними методами життєвого циклу](#identifying-unsafe-lifecycles)

> * [Попередженні про використання застарілого API строкових рефів](#warning-about-legacy-string-ref-api-usage)

Reviewer: строкових => рядкових

* [Попередженні про використання застарілого виклику findDOMNode](#warning-about-deprecated-finddomnode-usage)
* [Виявленні несподіваних побічних ефектів](#detecting-unexpected-side-effects)
* [Виявленні застарілого контекстного API](#detecting-legacy-context-api)

Додаткова функціональність буде додана в майбутніх релізах React.

### Ідентифікація небезпечних методів життєвого циклу {#identifying-unsafe-lifecycles}

[В цій статті](/blog/2018/03/27/update-on-async-rendering.html) пояснюється, чому деякі застарілі методи життєвого циклу є небезбечними для використання в асинхронних React-додатках. Проте може бути важко проконтролювати використання цих методів життєвого циклу, якщо ваш додаток використовує сторонні бібліотеки. На щастя, суворий режим може допомогти в цьому випадку!

Коли увімкнено суворий режим, то React створює список всіх класових компонентів, які використовують небезпечні методи життєвого циклу, та виводить попередження з інформацією про ці компоненти, наприклад:

![](../images/blog/strict-mode-unsafe-lifecycles-warning.png)

> Вирішення проблем, знайдених суворим режимом _сьогодні_ спростить для вас отримання вигоди від асинхронного рендеру в майбутніх релізах React.

Reviewer: `знайдених суворим режимом,` (кома)

### Попередження про використання застарілого API строкових рефів {#warning-about-legacy-string-ref-api-usage}

> Раніше React надавав два способи управління рефами: застарілий API строкових рефів та API функцій зворотнього виклику. Незважаючи на те, що API строкових рефів був зручніший у використанні, він мав [декілька недоліків](https://github.com/facebook/react/issues/1373) і тому ми офіційно рекомендували [використовувати форму функції зворотнього виклику](/docs/refs-and-the-dom.html#legacy-api-string-refs).

Reviewer: Строка [має бути](https://developer.mozilla.org/uk/docs/Web/JavaScript/Reference/Global_Objects#%D0%9E%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B0_%D1%82%D0%B5%D0%BA%D1%81%D1%82%D1%83) `рядок` по всьому документу.

> Раніше React надавав два способи управління рефами: застарілий API строкових рефів та API функцій зворотнього виклику. Незважаючи на те, що API строкових рефів був більш зручний в використанні, він мав [декілька недоліків](https://github.com/facebook/react/issues/1373) і тому ми офіційно рекомендували [використовувати форму функції зворотнього виклику](/docs/refs-and-the-dom.html#legacy-api-string-refs).

Reviewer 2: ```suggestion
Раніше React надавав два способи управління рефами: застарілий API строкових рефів та API функцій зворотнього виклику. Незважаючи на те, що API строкових рефів був зручніший у використанні, він мав [декілька недоліків](https://github.com/facebook/react/issues/1373) і тому ми офіційно рекомендували [використовувати форму функції зворотнього виклику](/docs/refs-and-the-dom.html#legacy-api-string-refs).
```

React версії 16.3 вводить третій варіант, який пропонує зручність строкових рефів але без їхніх недоліків:
`embed:16-3-release-blog-post/create-ref-example.js`

Оскільки об'єкти-рефи були додані значною мірою як заміна строкових рефів, то суворий режим тепер попереджує про використання строкових рефів.

> **Примітка:**
>

> > Рефи-функції зворотнього виклику продовжують підтримуватись на додачу до нового API-виклику `createRef`.

Reviewer: API-виклику => просто `API`?

>
> Вам не потрібно заміняти рефи-функції зворотнього виклику у ваших компонентах. Вони трохи більш гнучкі і тому залишаться як просунута можливість.

> [Дізнатись більше про новий API-виклик `createRef` можна тут.](/docs/refs-and-the-dom.html)

Reviewer: теж просто `API`

### Попередження про використання застарілого виклику findDOMNode {#warning-about-deprecated-finddomnode-usage}

> Колись React підтримував виклик `findDOMNode` для пошуку DOM-вузла в дереві по вказаному екзепляру класу. Зазвичай вам це не потрібно, тому що ви можете [прикріпити реф безпосередньо до DOM-вузла](/docs/refs-and-the-dom.html#creating-refs).

Reviewer 2: ```suggestion
Колись React підтримував виклик `findDOMNode` для пошуку DOM-вузла в дереві по вказаному екзепляру класа. Зазвичай вам це не потрібно, тому що ви можете [прикріпити реф безпосередньо до DOM-вузла](/docs/refs-and-the-dom.html#creating-refs).
```

> `findDOMNode` також може бути примінений до класових компонентів, але це порушувало рівні абстракції, дозволяючи батьківському компоненту вимагати щоб відбувся рендер певного дочірнього елементу. Це створює небезпеку рефакторингу, коли ви не можете змінити деталі реалізації компонента тому що батьківський компонент може використовувати його DOM-вузол. `findDOMNode` повертає лише перший дочірній елемент, але з використанням фрагментів компонент може рендерити декілька DOM-вузлів. `findDOMNode` — це API одноразового читання, він повертає вам результат лише на момент, коли ви його викликаєте. Якщо дочірній компонент відрендерить інший вузол то немає жодної можливості опрацювати цю зміну. Внаслідок цього `findDOMNode` працював лише якщо компоненти завжди повертали єдиний DOM-вузол що ніколи не змінювався.

Reviewer: примінений => `застосований` і кілька ком
```suggestion
`findDOMNode` також може бути застосований до класових компонентів, але це порушувало рівні абстракції, дозволяючи батьківському компоненту вимагати, щоб відбувся рендер певного дочірнього елементу. Це створює небезпеку рефакторингу, коли ви не можете змінити деталі реалізації компонента, тому що батьківський компонент може використовувати його DOM-вузол. `findDOMNode` повертає лише перший дочірній елемент, але з використанням фрагментів компонент може рендерити декілька DOM-вузлів. `findDOMNode` — це API одноразового читання, він повертає вам результат лише на момент, коли ви його викликаєте. Якщо дочірній компонент відрендерить інший вузол, то немає жодної можливості опрацювати цю зміну. Внаслідок цього `findDOMNode` працював лише якщо компоненти завжди повертали єдиний DOM-вузол, що ніколи не змінювався.
```

}
```

> Примітка:
>

> > Можна використати CSS атрибут [`display: contents`](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents) якщо ви не хочете, щоб вузол був частиною розмітки.

Reviewer 2: ```suggestion
> Можна використати CSS-атрибут [`display: contents`](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents) якщо ви не хочете, щоб вузол був частиною розмітки.
```

> > Можна використати CSS-атрибут [`display: contents`](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents) якщо ви не хочете, щоб вузол був частиною розмітки.

Reviewer: ```suggestion
> Можна використати CSS-атрибут [`display: contents`](https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents), якщо ви не хочете, щоб вузол був частиною розмітки.
```

### Виявлення несподіваних побічних ефектів {#detecting-unexpected-side-effects}

По суті, React виконує роботу в два етапи:
* Стадія **рендеру** визначає, які саме зміни потрібно зробити, наприклад в DOM. На цій стадії React викликає метод `render` і потім порівнює результат з попереднім рендером.
* Стадія **фіксації** — це коли React застосовує будь-які зміни. У випадку React DOM — це стадія, коли React вставляє, оновлює та видаляє вузли DOM. На цій стадії React також викликає методи життєвого циклу, такі як `componentDidMount` та `componentDidUpdate`.

> Стадія фіксації зазвичай дуже швидка, але рендер може бути повільним. З цієї причини майбутній асинхронний режим (який поки ще не увімкнений за замовчуванням) розбиває обсяг роботи рендеру на частини, призупиняючись і продовжуючи роботу щоб запобігти блокуванню браузера. Це означає, що React перед фіксацією може викликати методи життєвого циклу стадії рендеру більше ніж один раз, або викликати їх та взагалі не зафіксувати зміни (через помилку або переривання вищого пріоритету).

Reviewer: кома перед `щоб запобігти`

Методи життєвого циклу стадії рендеру включають наступні методи класового компоненту:
* `constructor`
* `componentWillMount`
* `componentWillReceiveProps`
* `componentWillUpdate`
* `getDerivedStateFromProps`
* `shouldComponentUpdate`
* `render`
* Функції-оновлювачі стану компонента, що передаються першим аргументом в `setState`

> Оскільки наведені вище методи можуть бути викликані більше ніж один раз, то важливо щоб вони не містили побічних ефектів. Ігнорування цього правила може привести до ряду проблем, включаючи витік пам'яті або недійсний стан додатку. На жаль виявити ці проблеми може бути важко оскільки вони часто бувають [недетермінованими](https://uk.wikipedia.org/wiki/Детермінований_алгоритм).

Reviewer: Кома перед `щоб`.
Кома після `На жаль`.
Кома перед `оскільки`.

> Оскільки наведені вище методи можуть бути викликані більше ніж один раз, то важливо щоб вони не містили побічних ефектів.  Ігнорування цього правила може привести до ряду проблем, включаючи витік пам'яті або недійсний стан додатку. На жаль виявити ці проблеми може бути важко оскільки вони часто бувають [недетермінованими](https://uk.wikipedia.org/wiki/Детермінований_алгоритм).

Author: зайвий пробіл перед `  Ігнорування`

Reviewer 2: Виправити 👆

> Суворий режим не може автоматично виявити побічні ефекти за вас але він може допомогти їх помітити, роблячи їх більш детермінованими. Це досягається завдяки навмисному викликлу наступних методів двічі:

Reviewer 2: ```suggestion
Суворий режим не може автоматично виявити побічні ефекти за вас, але він може допомогти їх помітити, роблячи їх більш детермінованими. Це досягається завдяки навмисному викликлу наступних методів двічі:
```

> * метод `constructor` класового компоненту

Reviewer 3: `метод` => `Метод`

Bullet lists should start with capital letters.

Author: Fixed, please take a look at this commit ac00a20

* Метод `render`
* Функції-оновлювачі стану компонента, що передаються першим аргументом в `setState`
* Статичний метод життєвого циклу `getDerivedStateFromProps`

> Примітка:
>
> Це застосовується лиже в режимі розробки. _Методи життєвого циклу ніколи не будуть викликані двічі в продакшн-режимі._

> Для прикладу розглянемо наступний код:

Reviewer: Кома після `Для прикладу`.

`embed:strict-mode/side-effects-in-constructor.js`

> 

Author: підскажіть як перекласти `subtle bug`?

Reviewer 2: Я б переклав, як `витончені помилки`. Начебто так можна казати про помилки. Знайшов таке використання в одній [статті на Вікіпедії](https://uk.wikipedia.org/wiki/%D0%9C%D0%B0%D1%82%D0%B5%D0%BC%D0%B0%D1%82%D0%B8%D1%87%D0%BD%D0%B8%D0%B9_%D1%81%D0%BE%D1%84%D1%96%D0%B7%D0%BC).

> На перший погляд він не здається проблематичним. Але якщо метод `SharedApplicationState.recordEvent` не [ідемпотентний](https://uk.wikipedia.org/wiki/Ідемпотентність), то створення багатьох екземплярів цього компонента може призвести до недійсного стану додатку. Такий тип витончених помилок може не проявляти себе під час розробки або робити це непослідовно, і тому може залишитися не поміченим.

Reviewer: Кома перед `і` зайва.

> Суворий режим робить подібні патерни більш помітними навмисно двічі викликаючи методи, такі як конструктор компонента.

Reviewer: кома перед `навмисно`.

### Виявлення застарілого контекстного API {#detecting-legacy-context-api}

> Використання застарілого контекстного API часто призводило до помилок, тому він буде видалений в майбутній мажорній версії React. Він все ще працює в усіх релізах версії 16.x але показуватиме це попередження у суворому режимі:

Reviewer 2: ```suggestion
Використання застарілого контекстного API часто призводило до помилок, тому він буде видалений в майбутній мажорній версії React. Він все ще працює в усіх релізах версії 16.x, але показуватиме це попередження у суворому режимі:
```
